### ISSUE TITLE
[BUG/ISSUE] GEOS-CHEM ERROR: Could not find DTRAIN in HEMCO data list

### ISSUE THREAD

**Issue by yuexuyan** [geoschem/geos-chem#258]
I ran a gobal merraCO2_2x2.5 simulation from 01/Jun/2015 to 31/May/2016 by GEOSChem 12.7.0.
It went smooth without any error or warnings  until I ran ./geos 
the GEOSChem stopped with this error message:
```
GEOS-CHEM ERROR: Could not find DTRAIN in HEMCO data list!
STOP at GET_MET_3D(get_met_mod.F90)
===============================================================================
     - CLEANUP: deallocating arrays now...
```
and when I open HEMCO.log file, there is also a warning saying: 
Data is treated as unitless, but file attribute suggests it is not...

here are my configuration files and the HEMCO.log
[HEMCO.log](https://github.com/geoschem/geos-chem/files/4388881/HEMCO.log)

**Comment by yuexuyan** [comment]
[HEMCO_Diagn.rc.txt](https://github.com/geoschem/geos-chem/files/4388959/HEMCO_Diagn.rc.txt)
[HISTORY.rc.txt](https://github.com/geoschem/geos-chem/files/4388960/HISTORY.rc.txt)
[input.geos.txt](https://github.com/geoschem/geos-chem/files/4388961/input.geos.txt)
[HEMCO_Config.rc.txt](https://github.com/geoschem/geos-chem/files/4388962/HEMCO_Config.rc.txt)

**Comment by yantosca** [comment]
In the HEMCO log the error is:
```
HEMCO ERROR: not enough time slices: DTRAIN
ERROR LOCATION: HCO_GetPtr_3D (hco_emislist_mod.F90)
```
Have you checked to make sure that you have all of the met fields for the period that you want to simulate?  Also check to make sure that the A3cld files are not corrupted.  Sometimes when you download data the file transfer can halt unexpectedly and leave a corrupted file.

**Comment by yuexuyan** [comment]
Thank you for the hint, I deleted A3cld file and re-downloaded it with dryrun option. 
Same error happened again

**Comment by yantosca** [comment]
Do you have the log.dryrun?

**Comment by yuexuyan** [comment]
[log.dryrun.unique.txt](https://github.com/geoschem/geos-chem/files/4389776/log.dryrun.unique.txt)
[log.dryrun.txt](https://github.com/geoschem/geos-chem/files/4389777/log.dryrun.txt)
yes, and there is another file generated in run directory

**Comment by yantosca** [comment]
Is your simulation "out-of-the-box" or did you make modifications?  You might try doing "make superclean" and then recompiling with the debugging flags to see if that picks up any errors.  I am wondering if you have an out of bounds error or something else that is causing this issue.

**Comment by yuexuyan** [comment]
I firstly ran a merra_CO2_2x2.5 simulation from Jun 2017 to May 2018 with some changes in Longitude min/max, Latitude  min/max and Number of levels in input.geos file:

%%% GRID MENU %%%       :
Grid resolution         : 2.0x2.5
Longitude min/max       : 100.0 180.0
Latitude  min/max       : -50.0  -5.0
 Half-sized polar boxes?: T
Number of levels        : 72
Nested grid simulation? : F 
 Buffer zone (N S E W ) :  0  0  0  0

and I also made changes in HISTORY.rc for the duration and frequency of restart collection and also CO2 collection:

restart collection
 Restart.frequency:          '00000100 000000',
  Restart.duration:           '00000100 000000',

CO2 collection
CO2.frequency:       00000001 000000
 CO2.duration:        00000100 000000

I just did the "make superclean" but same error happened after recompiling.

**Comment by yuexuyan** [comment]
after I change longitude and latitude back to default setting, error still happens

**Comment by lizziel** [comment]
DTRAIN is in the A3dyn file. Have you tried re-downloading the A3dyn file for that day?

**Comment by yuexuyan** [comment]
thank you Lizziel, I did that

**Comment by lizziel** [comment]
And it worked?

**Comment by yuexuyan** [comment]
no it doesn't

**Comment by yuexuyan** [comment]
I think I see what went wrong
it might relvent to a segmentation fault (invalid memory) but wasn't warned in error message.

I've set a breakpoint with GNU gdb at 'get_met_mod::get_met_3' and ran the GEOSChem, here are my command and message feedback:

**(gdb) info locals**
am_i_root = .FALSE.
fnd = .FALSE.
ptr3d = <error reading variable ptr3d (value requires 3985881888 bytes, which is more than max-value-size)>
rc = 120371976
t = 0
**(gdb) watch ptr3d** 
Hardware watchpoint 2: ptr3d
**(gdb) d 1**
**(gdb) c**
Continuing.
Warning:
Could not insert hardware watchpoint 2.
Could not insert hardware breakpoints:
You may have requested too many hardware breakpoints/watchpoints.

Command aborted.

I guess the variable ptr3d is the problem. However, the watch command seems not working on this issue. How to fix this variable?

**Comment by yuexuyan** [comment]
I think I see what went wrong
it might relvent to a segmentation fault (invalid memory) but wasn't warned in error message.
I've set a breakpoint with GNU gdb at 'get_met_mod::get_met_3d' and ran the GEOSChem, here are my command and message feedback:
(gdb) info locals
am_i_root = .FALSE.
fnd = .FALSE.
ptr3d = <error reading variable ptr3d (value requires 3985881888 bytes, which is more than max-value-size)>
rc = 120371976
t = 0
(gdb) watch ptr3d
Hardware watchpoint 2: ptr3d
(gdb) d 1
(gdb) c
Continuing.
Warning:
Could not insert hardware watchpoint 2.
Could not insert hardware breakpoints:
You may have requested too many hardware breakpoints/watchpoints.
Command aborted.
I guess the variable ptr3d is the problem. However, the watch command seems not working on this issue. How to fix this variable?

**Comment by yuexuyan** [comment]
I think I see what went wrong
it might relvent to a segmentation fault (invalid memory) but wasn't warned in error message.
I've set a breakpoint with GNU gdb at 'get_met_mod::get_met_3d' and ran the GEOSChem, here are my command and message feedback:

**(gdb) info locals**
am_i_root = .FALSE.
fnd = .FALSE.
ptr3d = <error reading variable ptr3d (value requires 3985881888 bytes, which is more than max-value-size)>
rc = 120371976
t = 0
**(gdb) watch ptr3d**
Hardware watchpoint 2: ptr3d
**(gdb) d 1**
**(gdb) c**
Continuing.
Warning:
Could not insert hardware watchpoint 2.
Could not insert hardware breakpoints:
You may have requested too many hardware breakpoints/watchpoints.
Command aborted.

I guess the variable ptr3d is the problem. However, the watch command seems not working on this issue. How to fix this variable?

**Comment by yantosca** [comment]
Try setting ```ulimit -n unlimited``` in your .bashrc and see if this works.

**Comment by yuexuyan** [comment]
Thank you Bob, I think it should work. But when I source .bashrc, the operation is not permitted. 
 I print ulimit information by "ulimit -a" it says:

core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 509755
max locked memory       (kbytes, -l) 16384
max memory size         (kbytes, -m) unlimited
open files                      (-n) 2048
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) unlimited
cpu time               (seconds, -t) unlimited
max user processes              (-u) 509755
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited

I've also tied to edit "/etc/security/limits.conf" file but I failed either. Maybe I need root permission to change it.

**Comment by yuexuyan** [comment]
Thank you Bob, I think it should work. But when I source .bashrc, the operation is not permitted.
I print ulimit information by "ulimit -a" it says:

core file size (blocks, -c) 0
data seg size (kbytes, -d) unlimited
scheduling priority (-e) 0
file size (blocks, -f) unlimited
pending signals (-i) 509755
max locked memory (kbytes, -l) 16384
max memory size (kbytes, -m) unlimited
open files (-n) 2048
pipe size (512 bytes, -p) 8
POSIX message queues (bytes, -q) 819200
real-time priority (-r) 0
stack size (kbytes, -s) unlimited
cpu time (seconds, -t) unlimited
max user processes (-u) 509755
virtual memory (kbytes, -v) unlimited
file locks (-x) unlimited

I've also tried to edit "/etc/security/limits.conf" file but I failed either. Maybe I need root permission to change it.

**Comment by yuexuyan** [comment]
oh, I find a way to set ulimit -n ulimited
now it is:
open files                      (-n)  1048576
but, this change doesn't work on my issue

**Comment by hgrhgy** [comment]
I have the same problem, is it solved now?

**Comment by hgrhgy** [comment]
the related code i found in the source: 
```fortran   !======================================================================
    ! Select the proper time slice
    !======================================================================

    ! Name of this routine (for error printout)
    caller  = "FlexGrid_Read_A3dyn (flexgrid_read_mod.F90)"

    ! Find the proper time-slice to read from disk
    t_index = ( HHMMSS / 030000 ) + 1 

    ! Stop w/ error if the time index is invalid
    IF ( t_index < 1 .or. t_index > 8 ) THEN
       WRITE( errMsg, 100 ) t_index
 100   FORMAT( 'Time_index value ', i5, ' must be in the range 1 to 8!' )
       CALL ERROR_STOP( errMsg, caller )
    ENDIF

    !======================================================================
    ! Get met fields from HEMCO
    !======================================================================

    ! Read DTRAIN
    v_name = "DTRAIN"
    CALL Get_Met_3D( State_Grid, Q, TRIM(v_name), t_index=t_index )
    State_Met%DTRAIN = Q
```
- function HCO_GetPtr_3D in hco_emlist_mod.F90 print the error msg
```fortran
...
   ! Check time dimension
    IF ( Lct%Dct%Dta%nt < T ) THEN
       MSG = 'not enough time slices: ' // TRIM(DctName)
       CALL HCO_ERROR( HcoState%Config%Err, MSG, RC, THISLOC=LOC )
       RETURN
    ENDIF
...
```

SO what the `Lct%Dct%Dta%nt ` stand for???

**Comment by yantosca** [comment]
Hi all -- I was able to replicate this issue.  It happens with debugging turned on or debugging turned off.  I suspect that the MERRA2 2 x2.5 A3dyn met data file for 20150601 @ 0430z might be corrupted.

@Jun-Meng, would you be able to re-run that day of data?

**Comment by Jun-Meng** [comment]
Hi all, 
I just re-run the data for that day. It should be on CC in a few minutes.

**Comment by hgrhgy** [comment]
I solved my problem by change `0-23` in the config file to `*`.  hope it helps。
```sh
# --- A3dyn fields ---
* DTRAIN    $METDIR/$YYYY/$MM/$MET.$YYYY$MM$DD.A3dyn.$RES.$NC  DTRAIN   1980-2019/1-12/1-31/0-23/+90minute RFY xyz 1  * -  1 1
* OMEGA     $METDIR/$YYYY/$MM/$MET.$YYYY$MM$DD.A3dyn.$RES.$NC  OMEGA    1980-2019/1-12/1-31/0-23/+90minute RFY xyz 1  * -  1 1
* RH        $METDIR/$YYYY/$MM/$MET.$YYYY$MM$DD.A3dyn.$RES.$NC  RH       1980-2019/1-12/1-31/0-23/+90minute RFY xyz 1  * -  1 1
* U         $METDIR/$YYYY/$MM/$MET.$YYYY$MM$DD.A3dyn.$RES.$NC  U        1980-2019/1-12/1-31/0-23/+90minute RFY xyz 1  * -  1 1
* V         $METDIR/$YYYY/$MM/$MET.$YYYY$MM$DD.A3dyn.$RES.$NC  V        1980-2019/1-12/1-31/0-23/+90minute RFY xyz 1  * -  1 1
```

**Comment by ltmurray** [comment]
BTW, I have this same problem with out-of-the-box 12.8.2 using GEOSFP 4x5 for July 2019. @hgrhgy's solution worked for me.

**Comment by yantosca** [comment]
I have confirmed this.  We will fix this in 12.9.2 in the HEMCO_Config.rc files.

**Comment by yantosca** [comment]
This issue has now been resolved by commit: https://github.com/geoschem/geos-chem-unittest/commit/b5c7638711bb2dc7c99a2519b12531f15572ea13.

NOTE: In GEOS-Chem 13.0.0, we will retire the geoschem/geos-chem-unitest repository and move run-directory creation to the geoschem/geos-chem repository.

**Comment by yantosca** [comment]
The corresponding bug fix for GEOS-Chem 13.0.0 has been pushed as commit https://github.com/geoschem/geos-chem/commit/5ff7d0d5d25a011bc542e11ba9a0147b35155d44.
